ip_proofs/src/gipa.rs

```diff
@@ -0,0 +1,556 @@
+                    (&m_a_recurse, &m_b_recurse),
+                    (&ck_a_recurse, &ck_b_recurse, ck_t),
+                    &c,
+                )?;
+
+                r_steps.push((com_1, com_2));
+                r_transcript.push(c);
+                Ok((r_steps, r_base, r_transcript, ck_base))
+            }
+            _ => unreachable!(), // If called only on message lengths power of 2
+        }
+    }
+
+    // Helper function used to calculate recursive challenges from proof execution (transcript in reverse)
+    pub fn verify_recursive_challenge_transcript(
+        com: (&LMC::Output, &RMC::Output, &IPC::Output),
+        proof: &GIPAProof<IP, LMC, RMC, IPC, D>,
+    ) -> Result<((LMC::Output, RMC::Output, IPC::Output), Vec<LMC::Scalar>), Error> {
+        let mut clone = Clone::clone(proof);
+        Self::_verify_recursive_challenges(com, &mut clone, &Default::default())
+    }
+
+    fn _verify_recursive_challenges(
+        com: (&LMC::Output, &RMC::Output, &IPC::Output),
+        proof: &mut GIPAProof<IP, LMC, RMC, IPC, D>,
+        transcript: &LMC::Scalar,
+    ) -> Result<((LMC::Output, RMC::Output, IPC::Output), Vec<LMC::Scalar>), Error> {
+        let (com_a, com_b, com_t) = com;
+        match proof.r_commitment_steps.len() {
+            0 => Ok(((com_a.clone(), com_b.clone(), com_t.clone()), Vec::new())), // base case
+            _ => {
+                // recursive step
+                // Fiat-Shamir challenge
+                let (com_1, com_2) = proof.r_commitment_steps.pop().unwrap();
+                let mut counter_nonce: usize = 0;
+                let (c, c_inv) = loop {
+                    let mut hash_input = Vec::new();
+                    hash_input.extend_from_slice(&counter_nonce.to_be_bytes()[..]);
+                    hash_input.extend_from_slice(&to_bytes![
+                        transcript, com_1.0, com_1.1, com_1.2, com_2.0, com_2.1, com_2.2
+                    ]?);
+                    if let Some(c) = LMC::Scalar::from_random_bytes(&D::digest(&hash_input)) {
+                        if let Some(c_inv) = c.inverse() {
+                            break (c, c_inv);
+                        }
+                    };
+                    counter_nonce += 1;
+                };
+
+                let com_recurse = (
+                    mul_helper(&com_1.0, &c) + com_a.clone() + mul_helper(&com_2.0, &c_inv),
+                    mul_helper(&com_1.1, &c) + com_b.clone() + mul_helper(&com_2.1, &c_inv),
+                    mul_helper(&com_1.2, &c) + com_t.clone() + mul_helper(&com_2.2, &c_inv),
+                );
+
+                let (base_com, mut challenges) = Self::_verify_recursive_challenges(
+                    (&com_recurse.0, &com_recurse.1, &com_recurse.2),
+                    proof,
+                    &c,
+                )?;
+                challenges.push(c);
+                Ok((base_com, challenges))
+            }
+        }
+    }
+
+    fn recursive_verify(
```

Comment:
Is there a way to reduce this to a single MSM, like what Bulletproofs does? Instead of essentially repeating the prover? This should provide great speed-ups for the verifier.

Reply:
Yes, we can do something like that (looking at Section 3.1 of [BBBPWM18](https://eprint.iacr.org/2017/1066.pdf)). 

I see that [BBBPWM18](https://eprint.iacr.org/2017/1066.pdf) describes the same bit representation trick that I had emailed you all about in the TIPP context to calculate these scalars in linear time (Section 6.2 "Computing scalars").

I changed the verification to perform the multiexponentiation all at once at the end. However, it does not currently get the gains of using `VariableMSM::multi_scalar_mul` since we do not bind the commitment keys (`LMC::Key`, `RMC::Key`) to `AffineCurve` -- a downside of the generalization. Perhaps, we can add two inner products to the generic parameterization to do this? Not a super clean solution... Thoughts?

Reply:
One way around this would be to add a "multi-scalar-mul" trait, and invoke that? That trait would have different implementations for fields vs for curves. (Maybe that belongs in zexe, and not here?)

Reply:
We can leave that to a future PR; I filed an issue for it.